pkg/kubelet/network/cni/cni.go

```diff
@@ -0,0 +1,232 @@
+	}
+	return networkMap
+}
+
+func (plugin *cniNetworkPlugin) selectNetwork(pod *api.Pod) (*cniNetwork, error) {
+	// loading networks all over again maybe inefficient, but required if networks can be created on the fly
+	plugin.cniNetworkMap = getAllCNINetworks()
+
+	if len(plugin.cniNetworkMap) == 0 {
+		return nil, fmt.Errorf("No available CNI network available in %s", DefaultNetDir)
+	}
+
+	// check if the namespace of the pod is a network name itself
+	network, ok := plugin.cniNetworkMap[pod.Namespace]
+	if ok {
+		return network, nil
+	}
+
+	// label called "network"?
+	network, ok = plugin.cniNetworkMap[pod.Labels["network"]]
+	if ok {
+		return network, nil
+	}
+
+	// annotation called "network"?
+	network, ok = plugin.cniNetworkMap[pod.Annotations["network"]]
+	if ok {
+		return network, nil
+	}
+
+	// else return the first one
+	for _, network := range plugin.cniNetworkMap {
+		return network, nil
+	}
+
+	return nil, fmt.Errorf("No available CNI network available in %s", DefaultNetDir)
+}
+
+func (plugin *cniNetworkPlugin) Init(host network.Host) error {
+	plugin.host = host
+	return nil
+}
+
+func (plugin *cniNetworkPlugin) Name() string {
+	return CNIPluginName
+}
+
+func (plugin *cniNetworkPlugin) SetUpPod(namespace string, name string, id kubeletTypes.DockerID) error {
+	netns, err := plugin.host.GetRuntime().GetNetNs(string(id))
+	if err != nil {
+		return err
+	}
+
+	pod, ok := plugin.host.GetPodByName(namespace, name)
+	if !ok {
+		return fmt.Errorf("pod %q namespace %q: unable to find pod", name, namespace)
+	}
+
+	// TODO: pick one network? all networks? which one writes to Status.PodIP?
+	// Perhaps pick the network by some annotation/label on pod, or the namespace.
+	network, err := plugin.selectNetwork(pod)
+	if err != nil {
+		return err
+	}
+	res, err := network.addNetwork(name, namespace, string(id), netns)
+
+	var ip string
+	if res.IP4 != nil {
+		ip = res.IP4.IP.String()
+	} else {
+		ip = res.IP6.IP.String()
+	}
+	// TODO-PAT: check that PodIP can be an IPv6.
+	pod.Status.PodIP = ip
+	return err
+}
+
+func (plugin *cniNetworkPlugin) TearDownPod(namespace string, name string, id kubeletTypes.DockerID) error {
+	netns, err := plugin.host.GetRuntime().GetNetNs(string(id))
+	if err != nil {
+		return err
+	}
+	pod, ok := plugin.host.GetPodByName(namespace, name)
+	if !ok {
+		return fmt.Errorf("pod %q namespace %q: unable to find pod", name, namespace)
+	}
+
+	network, err := plugin.selectNetwork(pod)
+	if err != nil {
+		return err
+	}
+	return network.deleteNetwork(name, namespace, string(id), netns)
+}
+
+func (plugin *cniNetworkPlugin) Status(namespace string, name string, id kubeletTypes.DockerID) (*network.PodNetworkStatus, error) {
+	pod, ok := plugin.host.GetPodByName(namespace, name)
+	if !ok {
+		return nil, fmt.Errorf("pod %q namespace %q: unable to find pod", name, namespace)
+	}
+
+	return &network.PodNetworkStatus{IP: net.ParseIP(pod.Status.PodIP)}, nil
```

Comment:
Paul,

That's a fair point.  I'll talk to @dchen1107 in the AM, but I suspect that
the risk of this is less bad than the cost of ad hoc checkpointing.

In the end there's a whole class of michief we simply CAN NOT PREVENT as
long as privileged containers are supported.

On Sun, Sep 20, 2015 at 11:21 AM, Paul Tiplady <notifications@github.com>
wrote:

> In pkg/kubelet/network/cni/cni.go
> <https://github.com/kubernetes/kubernetes/pull/13766#discussion_r39932742>
> :
>
> > +	}
> > +
> > +	network, err := plugin.selectNetwork(pod)
> > +	if err != nil {
> > +		return err
> > +	}
> > +	return network.deleteNetwork(name, namespace, string(id), netns)
> > +}
> > +
> > +func (plugin *cniNetworkPlugin) Status(namespace string, name string, id kubeletTypes.DockerID) (*network.PodNetworkStatus, error) {
> > +	pod, ok := plugin.host.GetPodByName(namespace, name)
> > +	if !ok {
> > +		return nil, fmt.Errorf("pod %q namespace %q: unable to find pod", name, namespace)
> > +	}
> > +
> > +	return &network.PodNetworkStatus{IP: net.ParseIP(pod.Status.PodIP)}, nil
>
> One possible issue with the nsenter approach: it changes the pod contract.
> Previously if a container changes its IP, docker (and therefore kubelet)
> will still report the original IP. That will probably break the pod, but
> nothing else. Now, we'll pick up the new IP and report that back to the
> cluster. What if a container changes its IP to clash with another pod's?
> What if the new IP is not in the host's cbr0 CIDR? Will that be handled
> gracefully? Not familiar enough with the k8s machinery involved to know if
> that's scary or not.
> … <#14febfd6fa0e01c8_>
> On Saturday, September 19, 2015, Tim Hockin <notifications@github.com>
> wrote: In pkg/kubelet/network/cni/cni.go <
> https://github.com/kubernetes/kubernetes/pull/13766#discussion_r39921151>
> : > + } > + > + network, err := plugin.selectNetwork(pod) > + if err != nil
> { > + return err > + } > + return network.deleteNetwork(name, namespace,
> string(id), netns) > +} > + > +func (plugin *cniNetworkPlugin)
> Status(namespace string, name string, id kubeletTypes.DockerID)
> (*network.PodNetworkStatus, error) { > + pod, ok :=
> plugin.host.GetPodByName(namespace, name) > + if !ok { > + return nil,
> fmt.Errorf("pod %q namespace %q: unable to find pod", name, namespace) > +
> } > + > + return &network.PodNetworkStatus{IP:
> net.ParseIP(pod.Status.PodIP)}, nil Stashing stuff on disk is the eventual
> answer, but we're trying REALLY hard to not introduce checkpointing in
> kubelet until we have time to do it carefully. We have burned our fingers a
> few times, and want to be cautious. Ad hoc checkpointing is the worst. IMO
> its better to read real-state for now even if it is a bit redundant. …
> <#14fe675098c24f0e_> On Sep 19, 2015 7:39 AM, "Paul Tiplady" <
> notifications@github.com <javascript:_e(%7B%7D,'cvml','
> notifications@github.com');>> wrote: In pkg/kubelet/network/cni/cni.go <
> https://github.com/kubernetes/kubernetes/pull/13766#discussion_r39920225>
> : > + } > + > + network, err := plugin.selectNetwork(pod) > + if err != nil
> { > + return err > + } > + return network.deleteNetwork(name, namespace,
> string(id), netns) > +} > + > +func (plugin *cniNetworkPlugin)
> Status(namespace string, name string, id kubeletTypes.DockerID)
> (*network.PodNetworkStatus, error) { > + pod, ok :=
> plugin.host.GetPodByName(namespace, name) > + if !ok { > + return nil,
> fmt.Errorf("pod %q namespace %q: unable to find pod", name, namespace) > +
> } > + > + return &network.PodNetworkStatus{IP:
> net.ParseIP(pod.Status.PodIP)}, nil Possibly hacky, but can you stash the
> IP on disk after initial pod creation (possibly at the path
> CNI_NAMESPACE/../IP)? This information is always localized to the node, and
> if the netns FD in that directory goes away then you'll have to recreate
> the whole pod anyway (including issuing a new IP). Or are we permitting the
> possibility that the IP might change? If we can avoid writing the nsenter
> code we'd gain on simplicity, would we lose anything? …
> <#14fe60bb5650329f_> On Friday, September 18, 2015, Tim Hockin <
> notifications@github.com <javascript:_e(%7B%7D,'cvml','
> notifications@github.com');>> wrote: In pkg/kubelet/network/cni/cni.go <
> https://github.com/kubernetes/kubernetes/pull/13766#discussion_r39916060>
> : > + } > + > + network, err := plugin.selectNetwork(pod) > + if err != nil
> { > + return err > + } > + return network.deleteNetwork(name, namespace,
> string(id), netns) > +} > + > +func (plugin *cniNetworkPlugin)
> Status(namespace string, name string, id kubeletTypes.DockerID)
> (*network.PodNetworkStatus, error) { > + pod, ok :=
> plugin.host.GetPodByName(namespace, name) > + if !ok { > + return nil,
> fmt.Errorf("pod %q namespace %q: unable to find pod", name, namespace) > +
> } > + > + return &network.PodNetworkStatus{IP:
> net.ParseIP(pod.Status.PodIP)}, nil In the mean time, we can nsenter, and
> have the discussion with @eyakubovich <https://github.com/eyakubovich> <
> https://github.com/eyakubovich> < https://github.com/eyakubovich> <
> https://github.com/eyakubovich> about offloading something to CNI — Reply
> to this email directly or view it on GitHub <
> https://github.com/kubernetes/kubernetes/pull/13766/files#r39916060>. —
> Reply to this email directly or view it on GitHub <
> https://github.com/kubernetes/kubernetes/pull/13766/files#r39920225>. —
> Reply to this email directly or view it on GitHub <
> https://github.com/kubernetes/kubernetes/pull/13766/files#r39921151>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13766/files#r39932742>.
>